### PR TITLE
feat(llamacpp): KV-cache prefix reuse across inference calls

### PR DESCRIPTION
**Pull request by kgentic-dev**
## Summary

- Add opt-in KV-cache prefix reuse to `LlamaCppProvider` via `context_reuse(true)`
- Persistent `SessionState` holds `LlamaContext` across calls — only delta tokens decoded
- Shared `acquire_context()` helper used by both chat and completion paths
- `ActiveContext` enum replaces dispatch macro for type-safe context access
- Public API: `reset_session()`, `cached_prefix_len()`
- 60 tests pass, 0 warnings

Currently `LlamaCppProvider` creates a fresh `LlamaContext` on every call — the full prompt is re-tokenized and decoded from scratch. For workloads with repeated system prompts (agents, multi-turn chat, RAG pipelines), this wastes significant prefill time.

With `context_reuse(true)`, the provider persists the context and compares each new prompt's tokens against the cached prefix. Only tokens that differ are decoded. The system prompt and tool definitions (often 100-500+ tokens) are decoded once and reused.

```rust
let provider = LlamaCppProvider::builder()
    .model_path("model.gguf")
    .context_reuse(true)
    .build()
    .await?;

// Call 1: full decode (cold start)
provider.chat_with_tools(&msgs1, Some(&tools), None).await?;

// Call 2: only new user message decoded, system+tools prefix reused
provider.chat_with_tools(&msgs2, Some(&tools), None).await?;

// Reset when context changes dramatically
provider.reset_session();
```

### Benchmark (Qwen3 0.6B, Apple Silicon)

| Call | Latency | Cached tokens |
|------|---------|---------------|
| Cold (no cache) | 93ms | 0 |
| Warm (prefix reuse) | 46ms | 36 |
| **Speedup** | **2.0x** | |

Tool-calling path caches 168 tokens (system + tool JSON). Larger models and longer system prompts see proportionally bigger gains.

### Safety

`SessionState` uses `unsafe { std::mem::transmute }` to erase the `LlamaContext<'a>` lifetime to `'static`. This is sound because:
1. The `Arc<LlamaModel>` that the context borrows from is stored in the same struct
2. Rust's field drop order guarantees the context drops before the model
3. A `Mutex` prevents concurrent access

This is the same pattern used by `llama-server`'s slot persistence and other production llama.cpp integrations.

### Design

- `acquire_context()` — shared by `generate_chat_text` and `generate_text`; handles lazy init, overflow reset, prefix comparison, and partial decode
- `ActiveContext` enum — type-safe dispatch between owned (fresh) and session (persistent) contexts via `with_ctx()` closure method
- `common_prefix_len()` — extracted free function, tested with 6 cases
- Overflow guard: if `next_pos + required_tokens > n_ctx`, context is reset automatically
- Generated tokens are evicted after each call — only prompt tokens cached

No behavior change when `context_reuse` is `false` (default).

## Test plan

- [ ] `cargo test --features metal` — 60 tests pass
- [ ] Config defaults: `context_reuse: false`, `enable_thinking: None`
- [ ] `common_prefix_len` — 6 edge cases (identical, diverges, shorter, empty, both_empty, different)
- [ ] Integration test with real GGUF model confirms 2x speedup (in consumer repo)
- [ ] `reset_session()` clears cache, next call re-initializes
- [ ] `cached_prefix_len()` returns 0 when disabled, >0 when active

### CODE REVIEW

File: crates/autoagents-llamacpp/src/provider.rs [hunk 1]
```diff
@@ -68,11 +68,184 @@ fn get_rt() -> &'static tokio::runtime::Runtime {
...
+            ))
+            .with_n_batch(n_batch);
+
+        let ctx_real = model
+            .new_context(&backend, ctx_params)
+            .map_err(|e| LlamaCppProviderError::ContextLoad(e.to_string()))?;
+
+        // SAFETY: see SessionState doc. model is kept alive in the same struct.
+        let ctx: llama_cpp_2::context::LlamaContext<'static> =
+            unsafe { std::mem::transmute(ctx_real) };
```
[the remark this answers is not in the document]
Comment (kgentic-dev) [reply crates/autoagents-llamacpp/src/provider.rs:146]: This is not T→T. The transmute is `LlamaContext<'a>` → `LlamaContext<'static>` — lifetime erasure, not type identity. Required because `LlamaContext` borrows from `&'a LlamaModel` at the FFI level, but we need to store both in the same struct (`SessionState`). The model is kept alive via `Arc<LlamaModel>` in the same struct, and field drop order guarantees the context drops first. See the `# Safety` doc on `SessionState`.

File: crates/autoagents-llamacpp/src/provider.rs [hunk 2]
```diff
@@ -1920,11 +2127,94 @@ fn build_chat_sampler(
...
+            }
+            ctx.decode(&mut batch)
+                .map_err(|err| LlamaCppProviderError::Inference(err.to_string()))?;
+            position += chunk.len() as i32;
+        }
+        Ok((ActiveContext::Owned(ctx), position))
+    }
+}
+
 fn generate_chat_text(
```
[the remark this answers is not in the document]
Comment (kgentic-dev) [reply crates/autoagents-llamacpp/src/provider.rs:2240]: Pre-existing complexity from the upstream `generate_chat_text` function. Our changes actually reduced it from 40 (PR #221 snapshot before extraction) to 26 by extracting `acquire_context()` as a shared helper. Further reduction would require restructuring the upstream generation loop which is out of scope for this feature PR.

File: crates/autoagents-llamacpp/src/provider.rs [hunk 3]
```diff
@@ -1920,11 +2127,94 @@ fn build_chat_sampler(
...
+                Arc::clone(model),
+                n_ctx,
+                n_batch,
+            )?);
+        }
+
+        {
+            let state = guard.as_mut().expect("session just initialised");
+
+            if state.next_pos as u32 + required_tokens > n_ctx {
```
[the remark this answers is not in the document]
Comment (kgentic-dev) [reply crates/autoagents-llamacpp/src/provider.rs]: Good catch — fixed in b41770b. Overflow check now uses `required_tokens > state.n_ctx` (the session's creation-time context window). Added `n_ctx` field to `SessionState` to track this. Previous logic would have caused unnecessary resets when successive prompts were moderately sized.

File: crates/autoagents-llamacpp/src/provider.rs [hunk 4]
```diff
@@ -1920,11 +2127,94 @@ fn build_chat_sampler(
...
+                state.evict_after(prefix_len)?;
+                let new_tokens = &prompt_tokens[prefix_len..];
+                if !new_tokens.is_empty() {
+                    state.decode_tokens(new_tokens, n_batch as usize)?;
+                }
+            }
+        }
+
+        let pos = guard.as_ref().expect("session exists").next_pos;
+        Ok((ActiveContext::Session(guard), pos))
```
[the remark this answers is not in the document]
Comment (kgentic-dev) [reply crates/autoagents-llamacpp/src/provider.rs:2212]: Acknowledged. Serialization is documented on the `context_reuse` config field (added in c70253a). Releasing the lock between prefill and generation would require splitting the context out of the MutexGuard, which adds complexity for a marginal gain — the target use case (single-agent sequential inference) doesn't benefit. Concurrent use should use separate provider instances.

File: crates/autoagents-llamacpp/src/provider.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kgentic-dev) [reply crates/autoagents-llamacpp/src/provider.rs:146]: Good point — added a doc comment in b41770b explaining why `Sync` is intentionally not implemented. `SessionState` must only be accessed through the `Mutex` wrapper; moving it into `Arc<SessionState>` directly would be UB since `LlamaContext` is not thread-safe.

**Comment by saivishwak** [comment]
@kgentic-dev , The core CI Issues Fix are merged in to main.

Thanks a lot for raising the PR.

**Comment by kgentic-dev** [comment]
Pushed fix for a bug discovered during integration testing:

**`3027746` — fix(llamacpp): refresh logits on 100% prefix match to prevent stale sampling**

When `context_reuse(true)` is enabled and consecutive calls use an identical prompt (e.g. user clicks "regenerate"), the prefix match is 100% and no new tokens are decoded in `acquire_context`. This left the logits buffer stale from the previous call's last generated token, causing the sampler to immediately hit EOG and produce 0 tokens.

Fix: when `new_tokens` is empty (full prefix hit), re-decode the last prompt token to refresh the logits buffer. Single-token decode — negligible cost.

Reproducer: call `chat_with_tools` twice with identical messages and `context_reuse(true)`. Without this fix, the second call returns empty.

All review findings from Greptile/DeepSource addressed in prior commits. CI should be green now that core fixes are merged to main.

**Comment by saivishwak** [comment]
Will Fix the Rust FMT and DeepSoruce Issues in another PR.
